React/forms.md

```diff
@@ -2,24 +2,280 @@
+
+We can add this to the constructor:
+
+```javascript
+//  new_student_form.js
+...
+constructor(props) {
+  this.state = {
+    name: '',
+    email: '',
+  }
+}
+```
+
+The `NewStudentForm` component will track the `name` and `email` as part of it's state.
+
+We can then manage the input components by setting the value of the form component to match the `NewStudentForm`'s state and adding an event handler to respond to changes in the input value.
+
+To turn the `name` input into a controlled component we first add an event handler:
+
+```javascript
+  //  new_student_form.js
+  onNameChange = (event) => {
+    console.log(`Name Field updated ${event.target.value}`)
+    this.setState({
+      name: event.target.value,
+    });
+  }
+```
+
+Then add `onChange` and `value` fields to the `input` in `render`.
+
+```javascript
+<input
+  onChange={ this.onNameChange }
+  value={ this.state.name }
+  name="name"
+/>
+```
+
+Now every time the user types into the name input field the `NewStudentForm`'s state is updated.
+
+**Practice**:  Update the e-mail field to have the `NewStudentForm` component manage it's state as well.
+
+**Question** Why would it be useful to track the state of a form field?
+
+
+## Form Validation
+
+By allowing the `NewStudentForm` component track the status of the form fields this will allow us to:
+- Validate form fields on the fly
+- More easily access the form fields when the form is submitted.
+
+We can perform a validation on the email field with a function like this:
+
+```javascript
+emailValid = () => {
+  return this.state.email.match(/\S+@\S+/);
+}
+```
+
+And give the user feedback on validation with:
+
+```javascript
+<input
+  onChange={ this.handleEmailChange }
+  value={ this.state.email }
+  className={ this.emailValid() ? "valid": "invalid" }
+  name="email"
+/>
+```
+
+**Question:**  What does this line with `className=` do?
+
+The form is rerendered every time the state of the component changes, and this code will run `this.emailValid()` and if the email field is valid the input will have a class of `valid`, and if not it will have the class of `invalid`.  With a little CSS we can give the user valuable feedback as to the status of a form field.
+
+```css
+.valid {
+  background: lightgreen;
+}
 
+.invalid {
+  background: pink;
+}
+```
 
+So with our form we can track input into the fields and provide real-time validation feedback.  Next we want to actually **do** something with the data, but our `NewStudentForm` component doesn't, and shouldn't need to care what happens with the submitted data.
+
+**Question**:  Looking at our Ada student list app to date, what component would be managing the student data?
+
+## Handling Submissions
+
+Now we want to handle when the user submits the form.  We can add a function as an event handler.
+
+```javascript
+// new_student_form.js
```

Comment:
The component naming convention will be camel case (i messed this up in some of my notes)